File: stage0/build-constants-plugins/src/main/java/net/rubygrapefruit/plugins/stage0/SettingsPlugin.java
package net.rubygrapefruit.plugins.stage0;

import org.gradle.api.Plugin;
import org.gradle.api.initialization.Settings;
import org.jetbrains.annotations.NotNull;

@SuppressWarnings("unused")
public class SettingsPlugin implements Plugin<Settings> {
    @Override
    public void apply(@NotNull Settings target) {
        target.getExtensions().create("buildConstants", BuildConstants.class);
    }
}

File: stage2/settings-plugins/src/main/kotlin/net/rubygrapefruit/plugins/stage2/SettingsPlugin.kt
package net.rubygrapefruit.plugins.stage2

import net.rubygrapefruit.plugins.stage0.BuildConstants
import org.gradle.api.Plugin
import org.gradle.api.initialization.Settings

@Suppress("unused")
class SettingsPlugin : Plugin<Settings> {
    override fun apply(target: Settings) {
        target.run {
            plugins.apply("org.gradle.toolchains.foojay-resolver-convention")
            gradle.rootProject { project ->
                project.run {
                    buildscript.repositories.mavenCentral()
                    buildscript.dependencies.add("classpath", BuildConstants.constants.kotlin.plugin.coordinates)

                    // For commonization
                    repositories.mavenCentral()
                }
            }
        }
    }
}